ambry-protocol/src/main/java/com.github.ambry.protocol/UndeleteResponse.java

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.protocol;
+
+import com.github.ambry.server.ServerErrorCode;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+
+
+/**
+ * Response of undelete rqeuest.
+ */
+public class UndeleteResponse extends Response {
+  private static final short UNDELETE_RESPONSE_VERSION_1 = 1;
+  public static final short INVALID_LIFE_VERSION = -1;
+
+  private short lifeVersion;
+  private static final int Life_Version_InBytes = Short.BYTES;
+
+  /**
+   * Constructs a {@link UndeleteResponse} with an {@link ServerErrorCode}.
+   * @param correlationId correlationId of the undelete response.
+   * @param clientId clientId of the undelete response.
+   * @param error error code returned in this undelete response.
+   */
+  public UndeleteResponse(int correlationId, String clientId, ServerErrorCode error) {
+    super(RequestOrResponseType.UndeleteResponse, UNDELETE_RESPONSE_VERSION_1, correlationId, clientId, error);
+    if (error == ServerErrorCode.No_Error) {
+      throw new IllegalArgumentException("NoError is not a valid error code");
```

Comment:
Can you explain a little more about this exception?

Reply:
there are two possible responses for undelete request: 1 Normal response with a lifeVersion and NO_ERROR code, 2 error response with a invalid lifeVersion number and any error code rather than NO_ERROR. That's why I am throwing out an exception here.